kernel/src/checkpoint/mod.rs

```diff
@@ -0,0 +1,485 @@
+        if !is_v2_checkpoint {
+            return Ok(None);
+        }
+        let values: &[Scalar] = &[version.into()];
+        // Create the nested schema structure for `CheckpointMetadata`
+        // Note: We cannot use `CheckpointMetadata::to_schema()` as it would include
+        // the 'tags' field which we're not supporting yet due to the lack of map support.
+        let schema = Arc::new(StructType::new([StructField::not_null(
+            "checkpointMetadata",
+            DataType::struct_type([StructField::not_null("version", DataType::LONG)]),
+        )]));
+
+        let checkpoint_metadata_batch = engine.evaluation_handler().create_one(schema, values)?;
+
+        let result = CheckpointData {
+            data: checkpoint_metadata_batch,
+            selection_vector: vec![true], // Always include this action
+        };
+
+        // Safe to mutably borrow counter here as the iterator has not yet been returned from
+        // `get_checkpoint_info()`. The iterator is the only other consumer of the counter.
+        let mut counter_ref = self
+            .total_actions_counter
+            .try_borrow_mut()
+            .map_err(|e| Error::generic(format!("Failed to borrow mutably: {}", e)))?;
+        *counter_ref += 1;
+
+        Ok(Some(Ok(result)))
+    }
+
+    /// Calculates the cutoff timestamp for deleted file cleanup.
+    ///
+    /// This function determines the minimum timestamp before which deleted files
+    /// will be permanently removed during VACUUM operations, based on the table's
+    /// deleted_file_retention_duration property.
+    ///
+    /// Returns the cutoff timestamp in milliseconds since epoch, matching
+    /// the remove action's deletion_timestamp format for comparison.
+    ///
+    /// The default retention period is 7 days, matching delta-spark's behavior.
+    fn deleted_file_retention_timestamp(&self) -> DeltaResult<i64> {
+        let retention_duration = self
+            .snapshot
+            .table_properties()
+            .deleted_file_retention_duration;
+
+        deleted_file_retention_timestamp_with_time(
+            retention_duration,
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .map_err(|e| Error::generic(format!("Failed to calculate system time: {}", e)))?,
+        )
+    }
+
+    /// Retrieves an iterator over all actions to be included in the checkpoint
+    ///
+    /// This method reads the relevant actions from the table's log segment using
+    /// the checkpoint schema, which filters for action types needed in checkpoints.
+    ///
+    /// The returned iterator yields tuples where:
+    /// - The first element is data in engine format
+    /// - The second element is a flag that indicates the action's source:
+    ///   - `true` if the action came from a commit file
+    ///   - `false` if the action came from a previous checkpoint file
+    fn replay_for_checkpoint_data(
```

Comment:
let's get away from using "replay" to describe "fetching the actions"
```suggestion
    fn read_checkpoint_actions(
```

Reply:
though this seems like a pretty trivial function - do we need it at all?

Reply:
Nope, I don't think we need it